File: 891-score-after-flipping-matrix/score-after-flipping-matrix.dart
import "dart:math";

class Solution {
  int matrixScore(List<List<int>> grid) {
    int m =grid.length,n=grid[0].length;
    for(int i = 0;i<m;i++) if(grid[i][0]==0) changeRow(grid[i]);
    int res = m * pow(2, n) as int;
    for(int i = 1 ; i < n ; i++){
      int oneCount = 0,zeroCount = 0;
      for(int j = 0 ; j < m ; j++) grid[j][i] == 1 ? oneCount++ : zeroCount++;
      res += max(oneCount,zeroCount) * pow(2,n-i) as int;
    }
    return res~/2;
  }
  void changeRow(List<int> nums) {for(var (i,v) in nums.indexed)nums[i] = v == 0 ? 1 : 0;}
}

File: 891-score-after-flipping-matrix/score-after-flipping-matrix.rs
impl Solution {
    fn matrix_score(mut grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let mut res = m as i32 * 2_i32.pow(n as u32);

        for i in 0..m {
            if grid[i][0] == 0 {
                Self::change_row(&mut grid[i]);
            }
        }


        for i in 1..n {
            let (mut one_count, mut zero_count) = (0, 0);
            for j in 0..m {
                if grid[j][i] == 1 {
                    one_count += 1;
                } else {
                    zero_count += 1;
                }
            }
            res += std::cmp::max(one_count, zero_count) * 2_i32.pow((n - i) as u32);
        }

        res / 2
    }

    fn change_row(nums: &mut Vec<i32>) {
        for num in nums.iter_mut() {
            if *num == 0 {
                *num = 1;
            } else {
                *num = 0;
            }
        }
    }
}